{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "db9f0bf883ae37d6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-21T12:37:18.562392Z",
     "start_time": "2025-08-21T12:37:17.439206Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"C:/Users/colin/PycharmProjects/PythonProject/tcga-brca-project/tcga-brca-vst-normalized-counts.csv\", index_col=0)\n",
    "metadata_df = pd.read_csv(\"C:/Users/colin/PycharmProjects/PythonProject/tcga-brca-project/tcga_brca_sample_metadata.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2d26e3433af7ec0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>patient</th>\n",
       "      <th>sample</th>\n",
       "      <th>shortLetterCode</th>\n",
       "      <th>definition</th>\n",
       "      <th>sample_submitter_id</th>\n",
       "      <th>sample_type_id</th>\n",
       "      <th>tumor_descriptor</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>...</th>\n",
       "      <th>paper_tobacco_smoking_history</th>\n",
       "      <th>paper_CNV.Clusters</th>\n",
       "      <th>paper_Mutation.Clusters</th>\n",
       "      <th>paper_DNA.Methylation.Clusters</th>\n",
       "      <th>paper_mRNA.Clusters</th>\n",
       "      <th>paper_miRNA.Clusters</th>\n",
       "      <th>paper_lncRNA.Clusters</th>\n",
       "      <th>paper_Protein.Clusters</th>\n",
       "      <th>paper_PARADIGM.Clusters</th>\n",
       "      <th>paper_Pan.Gyn.Clusters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-GM-A2DL-01A-11R-A18M-07</th>\n",
       "      <td>TCGA-GM-A2DL-01A-11R-A18M-07</td>\n",
       "      <td>TCGA-GM-A2DL</td>\n",
       "      <td>TCGA-GM-A2DL-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-GM-A2DL-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary</td>\n",
       "      <td>d6e981ee-1afc-4f84-a3be-e87e0d1460cd</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>C5</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AC-A2QI-01A-12R-A19W-07</th>\n",
       "      <td>TCGA-AC-A2QI-01A-12R-A19W-07</td>\n",
       "      <td>TCGA-AC-A2QI</td>\n",
       "      <td>TCGA-AC-A2QI-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-AC-A2QI-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary</td>\n",
       "      <td>08f10bb1-6dd4-4f36-86a3-1801f9762187</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C1</td>\n",
       "      <td>C4</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>C2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C1</td>\n",
       "      <td>C6</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-EW-A1PD-01A-11R-A144-07</th>\n",
       "      <td>TCGA-EW-A1PD-01A-11R-A144-07</td>\n",
       "      <td>TCGA-EW-A1PD</td>\n",
       "      <td>TCGA-EW-A1PD-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-EW-A1PD-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary</td>\n",
       "      <td>9f52ae8b-7fe7-441c-9fb6-c55a9bd1a0d0</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AR-A24N-01A-11R-A169-07</th>\n",
       "      <td>TCGA-AR-A24N-01A-11R-A169-07</td>\n",
       "      <td>TCGA-AR-A24N</td>\n",
       "      <td>TCGA-AR-A24N-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-AR-A24N-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary</td>\n",
       "      <td>dcc31bf1-9890-4fb1-9677-b5fb77e1fc37</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C1</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>C1</td>\n",
       "      <td>C3</td>\n",
       "      <td>C2</td>\n",
       "      <td>C1</td>\n",
       "      <td>C5</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AR-A24U-01A-11R-A169-07</th>\n",
       "      <td>TCGA-AR-A24U-01A-11R-A169-07</td>\n",
       "      <td>TCGA-AR-A24U</td>\n",
       "      <td>TCGA-AR-A24U-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-AR-A24U-01A</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary</td>\n",
       "      <td>17cdb451-9b35-4aab-8bc2-b4a64ef3e3b4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C4</td>\n",
       "      <td>C4</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>C3</td>\n",
       "      <td>C1</td>\n",
       "      <td>C1</td>\n",
       "      <td>C4</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 96 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   barcode       patient  \\\n",
       "TCGA-GM-A2DL-01A-11R-A18M-07  TCGA-GM-A2DL-01A-11R-A18M-07  TCGA-GM-A2DL   \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07  TCGA-AC-A2QI-01A-12R-A19W-07  TCGA-AC-A2QI   \n",
       "TCGA-EW-A1PD-01A-11R-A144-07  TCGA-EW-A1PD-01A-11R-A144-07  TCGA-EW-A1PD   \n",
       "TCGA-AR-A24N-01A-11R-A169-07  TCGA-AR-A24N-01A-11R-A169-07  TCGA-AR-A24N   \n",
       "TCGA-AR-A24U-01A-11R-A169-07  TCGA-AR-A24U-01A-11R-A169-07  TCGA-AR-A24U   \n",
       "\n",
       "                                        sample shortLetterCode  \\\n",
       "TCGA-GM-A2DL-01A-11R-A18M-07  TCGA-GM-A2DL-01A              TP   \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07  TCGA-AC-A2QI-01A              TP   \n",
       "TCGA-EW-A1PD-01A-11R-A144-07  TCGA-EW-A1PD-01A              TP   \n",
       "TCGA-AR-A24N-01A-11R-A169-07  TCGA-AR-A24N-01A              TP   \n",
       "TCGA-AR-A24U-01A-11R-A169-07  TCGA-AR-A24U-01A              TP   \n",
       "\n",
       "                                       definition sample_submitter_id  \\\n",
       "TCGA-GM-A2DL-01A-11R-A18M-07  Primary solid Tumor    TCGA-GM-A2DL-01A   \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07  Primary solid Tumor    TCGA-AC-A2QI-01A   \n",
       "TCGA-EW-A1PD-01A-11R-A144-07  Primary solid Tumor    TCGA-EW-A1PD-01A   \n",
       "TCGA-AR-A24N-01A-11R-A169-07  Primary solid Tumor    TCGA-AR-A24N-01A   \n",
       "TCGA-AR-A24U-01A-11R-A169-07  Primary solid Tumor    TCGA-AR-A24U-01A   \n",
       "\n",
       "                              sample_type_id tumor_descriptor  \\\n",
       "TCGA-GM-A2DL-01A-11R-A18M-07               1          Primary   \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07               1          Primary   \n",
       "TCGA-EW-A1PD-01A-11R-A144-07               1          Primary   \n",
       "TCGA-AR-A24N-01A-11R-A169-07               1          Primary   \n",
       "TCGA-AR-A24U-01A-11R-A169-07               1          Primary   \n",
       "\n",
       "                                                         sample_id  \\\n",
       "TCGA-GM-A2DL-01A-11R-A18M-07  d6e981ee-1afc-4f84-a3be-e87e0d1460cd   \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07  08f10bb1-6dd4-4f36-86a3-1801f9762187   \n",
       "TCGA-EW-A1PD-01A-11R-A144-07  9f52ae8b-7fe7-441c-9fb6-c55a9bd1a0d0   \n",
       "TCGA-AR-A24N-01A-11R-A169-07  dcc31bf1-9890-4fb1-9677-b5fb77e1fc37   \n",
       "TCGA-AR-A24U-01A-11R-A169-07  17cdb451-9b35-4aab-8bc2-b4a64ef3e3b4   \n",
       "\n",
       "                                sample_type  ...  \\\n",
       "TCGA-GM-A2DL-01A-11R-A18M-07  Primary Tumor  ...   \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07  Primary Tumor  ...   \n",
       "TCGA-EW-A1PD-01A-11R-A144-07  Primary Tumor  ...   \n",
       "TCGA-AR-A24N-01A-11R-A169-07  Primary Tumor  ...   \n",
       "TCGA-AR-A24U-01A-11R-A169-07  Primary Tumor  ...   \n",
       "\n",
       "                             paper_tobacco_smoking_history  \\\n",
       "TCGA-GM-A2DL-01A-11R-A18M-07                           NaN   \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07                           NaN   \n",
       "TCGA-EW-A1PD-01A-11R-A144-07                           NaN   \n",
       "TCGA-AR-A24N-01A-11R-A169-07                           NaN   \n",
       "TCGA-AR-A24U-01A-11R-A169-07                           NaN   \n",
       "\n",
       "                              paper_CNV.Clusters paper_Mutation.Clusters  \\\n",
       "TCGA-GM-A2DL-01A-11R-A18M-07                  C3                      C1   \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07                  C1                      C4   \n",
       "TCGA-EW-A1PD-01A-11R-A144-07                 NaN                     NaN   \n",
       "TCGA-AR-A24N-01A-11R-A169-07                  C1                      C1   \n",
       "TCGA-AR-A24U-01A-11R-A169-07                  C4                      C4   \n",
       "\n",
       "                              paper_DNA.Methylation.Clusters  \\\n",
       "TCGA-GM-A2DL-01A-11R-A18M-07                              C2   \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07                              C1   \n",
       "TCGA-EW-A1PD-01A-11R-A144-07                             NaN   \n",
       "TCGA-AR-A24N-01A-11R-A169-07                              C2   \n",
       "TCGA-AR-A24U-01A-11R-A169-07                              C1   \n",
       "\n",
       "                             paper_mRNA.Clusters paper_miRNA.Clusters  \\\n",
       "TCGA-GM-A2DL-01A-11R-A18M-07                  C1                   C2   \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07                  C2                   C2   \n",
       "TCGA-EW-A1PD-01A-11R-A144-07                 NaN                  NaN   \n",
       "TCGA-AR-A24N-01A-11R-A169-07                  C1                   C3   \n",
       "TCGA-AR-A24U-01A-11R-A169-07                  C2                   C3   \n",
       "\n",
       "                             paper_lncRNA.Clusters  paper_Protein.Clusters  \\\n",
       "TCGA-GM-A2DL-01A-11R-A18M-07                    C1                      C2   \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07                   NaN                      C1   \n",
       "TCGA-EW-A1PD-01A-11R-A144-07                   NaN                     NaN   \n",
       "TCGA-AR-A24N-01A-11R-A169-07                    C2                      C1   \n",
       "TCGA-AR-A24U-01A-11R-A169-07                    C1                      C1   \n",
       "\n",
       "                             paper_PARADIGM.Clusters  paper_Pan.Gyn.Clusters  \n",
       "TCGA-GM-A2DL-01A-11R-A18M-07                      C5                      C3  \n",
       "TCGA-AC-A2QI-01A-12R-A19W-07                      C6                      C1  \n",
       "TCGA-EW-A1PD-01A-11R-A144-07                     NaN                     NaN  \n",
       "TCGA-AR-A24N-01A-11R-A169-07                      C5                      C1  \n",
       "TCGA-AR-A24U-01A-11R-A169-07                      C4                      C3  \n",
       "\n",
       "[5 rows x 96 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()\n",
    "metadata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "87a4c1048adf48f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import train_test_split\n",
    "import torch\n",
    "from torch.utils.data import DataLoader, TensorDataset\n",
    "import numpy as np\n",
    "\n",
    "X = df.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3091ed66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Expression Data Shape (Samples x Genes): (885, 18303)\n",
      "Metadata Shape (Samples x Features): (885, 96)\n",
      "Are all sample IDs aligned? True\n"
     ]
    }
   ],
   "source": [
    "common_samples = list(set(X.index) & set(metadata_df.index))\n",
    "X = X.loc[common_samples].sort_index()\n",
    "\n",
    "meta_df = metadata_df.loc[common_samples].sort_index()\n",
    "\n",
    "print(f\"\\nExpression Data Shape (Samples x Genes): {X.shape}\")\n",
    "print(f\"Metadata Shape (Samples x Features): {meta_df.shape}\")\n",
    "print(f\"Are all sample IDs aligned? {all(X.index == meta_df.index)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b386d0fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Aligned Expression Data (First 5 Sample IDs):\n",
      "['TCGA-A1-A0SB-01A-11R-A144-07', 'TCGA-A1-A0SF-01A-11R-A144-07', 'TCGA-A1-A0SG-01A-11R-A144-07', 'TCGA-A1-A0SI-01A-11R-A144-07', 'TCGA-A1-A0SN-01A-11R-A144-07']\n",
      "\n",
      "Aligned Metadata (First 5 Sample IDs):\n",
      "['TCGA-A1-A0SB-01A-11R-A144-07', 'TCGA-A1-A0SF-01A-11R-A144-07', 'TCGA-A1-A0SG-01A-11R-A144-07', 'TCGA-A1-A0SI-01A-11R-A144-07', 'TCGA-A1-A0SN-01A-11R-A144-07']\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nAligned Expression Data (First 5 Sample IDs):\")\n",
    "print(X.index[:5].tolist())\n",
    "print(\"\\nAligned Metadata (First 5 Sample IDs):\")\n",
    "print(meta_df.index[:5].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fab4511a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Scaled Expression Data (Head):\n",
      "                              ENSG00000000003.15  ENSG00000000419.13  \\\n",
      "TCGA-A1-A0SB-01A-11R-A144-07            0.901053           -1.107678   \n",
      "TCGA-A1-A0SF-01A-11R-A144-07            1.964514           -0.397228   \n",
      "TCGA-A1-A0SG-01A-11R-A144-07            0.225169           -0.483863   \n",
      "TCGA-A1-A0SI-01A-11R-A144-07           -0.125770            0.258707   \n",
      "TCGA-A1-A0SN-01A-11R-A144-07           -1.736456            2.242947   \n",
      "\n",
      "                              ENSG00000000457.14  ENSG00000000460.17  \\\n",
      "TCGA-A1-A0SB-01A-11R-A144-07           -0.537444           -0.635950   \n",
      "TCGA-A1-A0SF-01A-11R-A144-07           -0.042405           -0.195591   \n",
      "TCGA-A1-A0SG-01A-11R-A144-07           -0.875888           -0.593551   \n",
      "TCGA-A1-A0SI-01A-11R-A144-07            1.146420            1.751878   \n",
      "TCGA-A1-A0SN-01A-11R-A144-07            0.415368            0.513456   \n",
      "\n",
      "                              ENSG00000000938.13  ENSG00000000971.16  \\\n",
      "TCGA-A1-A0SB-01A-11R-A144-07           -1.712165            1.641017   \n",
      "TCGA-A1-A0SF-01A-11R-A144-07            0.495801            0.193284   \n",
      "TCGA-A1-A0SG-01A-11R-A144-07           -0.288065            0.206657   \n",
      "TCGA-A1-A0SI-01A-11R-A144-07           -0.484578            1.066979   \n",
      "TCGA-A1-A0SN-01A-11R-A144-07           -0.281046            0.466929   \n",
      "\n",
      "                              ENSG00000001036.14  ENSG00000001084.13  \\\n",
      "TCGA-A1-A0SB-01A-11R-A144-07           -1.947701           -0.473659   \n",
      "TCGA-A1-A0SF-01A-11R-A144-07           -0.783097           -0.547196   \n",
      "TCGA-A1-A0SG-01A-11R-A144-07           -0.147475            0.045047   \n",
      "TCGA-A1-A0SI-01A-11R-A144-07            0.423888           -1.122231   \n",
      "TCGA-A1-A0SN-01A-11R-A144-07            0.813513           -0.941770   \n",
      "\n",
      "                              ENSG00000001167.14  ENSG00000001460.18  ...  \\\n",
      "TCGA-A1-A0SB-01A-11R-A144-07           -0.056775            1.028051  ...   \n",
      "TCGA-A1-A0SF-01A-11R-A144-07            0.310305            0.726930  ...   \n",
      "TCGA-A1-A0SG-01A-11R-A144-07           -0.291072            0.944642  ...   \n",
      "TCGA-A1-A0SI-01A-11R-A144-07            0.275946            0.495363  ...   \n",
      "TCGA-A1-A0SN-01A-11R-A144-07           -0.541491           -0.964228  ...   \n",
      "\n",
      "                              ENSG00000288559.1  ENSG00000288573.1  \\\n",
      "TCGA-A1-A0SB-01A-11R-A144-07          -0.307253           0.485651   \n",
      "TCGA-A1-A0SF-01A-11R-A144-07           0.813083           1.046818   \n",
      "TCGA-A1-A0SG-01A-11R-A144-07          -0.157356           0.362826   \n",
      "TCGA-A1-A0SI-01A-11R-A144-07           0.369557          -0.234208   \n",
      "TCGA-A1-A0SN-01A-11R-A144-07           0.385831          -1.254426   \n",
      "\n",
      "                              ENSG00000288585.1  ENSG00000288586.1  \\\n",
      "TCGA-A1-A0SB-01A-11R-A144-07          -0.597459           0.475821   \n",
      "TCGA-A1-A0SF-01A-11R-A144-07          -0.342963          -0.893049   \n",
      "TCGA-A1-A0SG-01A-11R-A144-07           1.450317           0.324510   \n",
      "TCGA-A1-A0SI-01A-11R-A144-07           0.039627           0.025419   \n",
      "TCGA-A1-A0SN-01A-11R-A144-07          -0.012789          -0.766188   \n",
      "\n",
      "                              ENSG00000288596.2  ENSG00000288598.1  \\\n",
      "TCGA-A1-A0SB-01A-11R-A144-07          -0.164475           1.015015   \n",
      "TCGA-A1-A0SF-01A-11R-A144-07          -0.998560           0.374208   \n",
      "TCGA-A1-A0SG-01A-11R-A144-07           0.821322           0.011711   \n",
      "TCGA-A1-A0SI-01A-11R-A144-07          -1.114945          -1.684796   \n",
      "TCGA-A1-A0SN-01A-11R-A144-07          -2.041750          -0.608637   \n",
      "\n",
      "                              ENSG00000288612.1  ENSG00000288663.1  \\\n",
      "TCGA-A1-A0SB-01A-11R-A144-07          -0.416598           0.161732   \n",
      "TCGA-A1-A0SF-01A-11R-A144-07          -0.737455          -0.252792   \n",
      "TCGA-A1-A0SG-01A-11R-A144-07          -0.634214          -0.565149   \n",
      "TCGA-A1-A0SI-01A-11R-A144-07          -0.289156          -0.683200   \n",
      "TCGA-A1-A0SN-01A-11R-A144-07          -2.201589          -1.362445   \n",
      "\n",
      "                              ENSG00000288670.1  ENSG00000288675.1  \n",
      "TCGA-A1-A0SB-01A-11R-A144-07          -0.823509           0.279546  \n",
      "TCGA-A1-A0SF-01A-11R-A144-07           0.568187           0.515681  \n",
      "TCGA-A1-A0SG-01A-11R-A144-07           0.507903           0.916734  \n",
      "TCGA-A1-A0SI-01A-11R-A144-07          -0.180042           0.558689  \n",
      "TCGA-A1-A0SN-01A-11R-A144-07          -0.633294          -0.684025  \n",
      "\n",
      "[5 rows x 18303 columns]\n"
     ]
    }
   ],
   "source": [
    "scaler = StandardScaler()\n",
    "X_scaled = pd.DataFrame(scaler.fit_transform(X), columns=X.columns, index=X.index)\n",
    "print(\"\\nScaled Expression Data (Head):\")\n",
    "print(X_scaled.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5832a94b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Target Variable Unique Values (paper_BRCA_Subtype_PAM50): ['Normal' 'LumA' 'LumB' 'Basal' 'Her2']\n",
      "Encoded Labels (First 5): [4 2 2 3 3]\n"
     ]
    }
   ],
   "source": [
    "if 'paper_BRCA_Subtype_PAM50' not in meta_df.columns:\n",
    "    print(\"Error: 'paper_BRCA_Subtype_PAM50' not found in metadata. Please choose an existing column.\")\n",
    "    if 'vital_status' in meta_df.columns:\n",
    "        target_variable = 'vital_status'\n",
    "        print(f\"Using '{target_variable}' as target variable instead.\")\n",
    "    else:\n",
    "        raise ValueError(\"No suitable target variable found for kNN classification.\")\n",
    "else:\n",
    "    target_variable = 'paper_BRCA_Subtype_PAM50'\n",
    "\n",
    "y_labels = meta_df[target_variable]\n",
    "\n",
    "valid_indices = y_labels.dropna().index\n",
    "X_knn = X_scaled.loc[valid_indices]\n",
    "y_knn = y_labels.loc[valid_indices]\n",
    "\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "label_encoder = LabelEncoder()\n",
    "y_encoded = label_encoder.fit_transform(y_knn)\n",
    "\n",
    "print(f\"\\nTarget Variable Unique Values ({target_variable}): {y_knn.unique()}\")\n",
    "print(f\"Encoded Labels (First 5): {y_encoded[:5]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "57e4ccaf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training set size: 605 samples\n",
      "Test set size: 152 samples\n"
     ]
    }
   ],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(\n",
    "    X_knn, y_encoded, test_size=0.2, random_state=42, stratify=y_encoded\n",
    ")\n",
    "\n",
    "print(f\"\\nTraining set size: {X_train.shape[0]} samples\")\n",
    "print(f\"Test set size: {X_test.shape[0]} samples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d2e6031c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.metrics import accuracy_score, classification_report\n",
    "\n",
    "knn = KNeighborsClassifier(n_neighbors=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "355c1692",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "kNN Classification Accuracy: 0.7566\n",
      "\n",
      "kNN Classification Report:\n",
      "               precision    recall  f1-score   support\n",
      "\n",
      "       Basal       0.96      0.90      0.93        29\n",
      "        Her2       0.67      0.20      0.31        10\n",
      "        LumA       0.74      0.98      0.84        82\n",
      "        LumB       0.83      0.20      0.32        25\n",
      "      Normal       0.25      0.33      0.29         6\n",
      "\n",
      "    accuracy                           0.76       152\n",
      "   macro avg       0.69      0.52      0.54       152\n",
      "weighted avg       0.77      0.76      0.72       152\n",
      "\n"
     ]
    }
   ],
   "source": [
    "knn.fit(X_train, y_train)\n",
    "\n",
    "y_pred = knn.predict(X_test)\n",
    "\n",
    "accuracy = accuracy_score(y_test, y_pred)\n",
    "report = classification_report(y_test, y_pred, target_names=label_encoder.classes_)\n",
    "print(f\"\\nkNN Classification Accuracy: {accuracy:.4f}\")\n",
    "print(\"\\nkNN Classification Report:\\n\", report)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6e3092b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Best k for kNN: 2\n",
      "Best cross-validation accuracy: 0.7623\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import GridSearchCV\n",
    "param_grid = {'n_neighbors': range(1, 31)}\n",
    "grid_search = GridSearchCV(KNeighborsClassifier(), param_grid, cv=5, scoring='accuracy')\n",
    "grid_search.fit(X_knn, y_encoded)\n",
    "\n",
    "print(f\"\\nBest k for kNN: {grid_search.best_params_['n_neighbors']}\")\n",
    "print(f\"Best cross-validation accuracy: {grid_search.best_score_:.4f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44dd864b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of features (genes): 18303\n",
      "Number of classes: 5\n"
     ]
    }
   ],
   "source": [
    "X_train_tensor = torch.tensor(X_train.values, dtype=torch.float32)\n",
    "y_train_tensor = torch.tensor(y_train, dtype=torch.long)\n",
    "\n",
    "X_test_tensor = torch.tensor(X_test.values, dtype=torch.float32)\n",
    "y_test_tensor = torch.tensor(y_test, dtype=torch.long)\n",
    "\n",
    "train_dataset = TensorDataset(X_train_tensor, y_train_tensor)\n",
    "test_dataset = TensorDataset(X_test_tensor, y_test_tensor)\n",
    "\n",
    "batch_size = 64\n",
    "train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)\n",
    "test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3e4b0b81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Neural Network Model Architecture:\n",
      "GeneExpressionClassifier(\n",
      "  (layer1): Linear(in_features=18303, out_features=256, bias=True)\n",
      "  (relu1): ReLU()\n",
      "  (dropout1): Dropout(p=0.3, inplace=False)\n",
      "  (layer2): Linear(in_features=256, out_features=128, bias=True)\n",
      "  (relu2): ReLU()\n",
      "  (dropout2): Dropout(p=0.3, inplace=False)\n",
      "  (output_layer): Linear(in_features=128, out_features=5, bias=True)\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "import torch.nn as nn\n",
    "import torch.optim as optim\n",
    "\n",
    "class GeneExpressionClassifier(nn.Module):\n",
    "    def __init__(self, input_size, num_classes):\n",
    "        super(GeneExpressionClassifier, self).__init__()\n",
    "        self.layer1 = nn.Linear(input_size, 256)\n",
    "        self.relu1 = nn.ReLU()\n",
    "        self.dropout1 = nn.Dropout(0.3)\n",
    "\n",
    "        self.layer2 = nn.Linear(256, 128)\n",
    "        self.relu2 = nn.ReLU()\n",
    "        self.dropout2 = nn.Dropout(0.3)\n",
    "\n",
    "        self.output_layer = nn.Linear(128, num_classes)\n",
    "\n",
    "    def forward(self, x):\n",
    "        x = self.layer1(x)\n",
    "        x = self.relu1(x)\n",
    "        x = self.dropout1(x)\n",
    "\n",
    "        x = self.layer2(x)\n",
    "        x = self.relu2(x)\n",
    "        x = self.dropout2(x)\n",
    "\n",
    "        x = self.output_layer(x)\n",
    "        return x\n",
    "\n",
    "input_size = X_train.shape[1]\n",
    "num_classes = len(label_encoder.classes_)\n",
    "model = GeneExpressionClassifier(input_size, num_classes)\n",
    "\n",
    "criterion = nn.CrossEntropyLoss()\n",
    "optimizer = optim.Adam(model.parameters(), lr=0.001)\n",
    "\n",
    "print(\"\\nNeural Network Model Architecture:\")\n",
    "print(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6e23ec88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training on: cuda\n",
      "Epoch 1/50, Loss: 0.8355\n",
      "Epoch 2/50, Loss: 0.4342\n",
      "Epoch 3/50, Loss: 0.3608\n",
      "Epoch 4/50, Loss: 0.3185\n",
      "Epoch 5/50, Loss: 0.1517\n",
      "Epoch 6/50, Loss: 0.1005\n",
      "Epoch 7/50, Loss: 0.0760\n",
      "Epoch 8/50, Loss: 0.0296\n",
      "Epoch 9/50, Loss: 0.0832\n",
      "Epoch 10/50, Loss: 0.0602\n",
      "Epoch 11/50, Loss: 0.0717\n",
      "Epoch 12/50, Loss: 0.0770\n",
      "Epoch 13/50, Loss: 0.0308\n",
      "Epoch 14/50, Loss: 0.0303\n",
      "Epoch 15/50, Loss: 0.0293\n",
      "Epoch 16/50, Loss: 0.0268\n",
      "Epoch 17/50, Loss: 0.0559\n",
      "Epoch 18/50, Loss: 0.0481\n",
      "Epoch 19/50, Loss: 0.0860\n",
      "Epoch 20/50, Loss: 0.0901\n",
      "Epoch 21/50, Loss: 0.0807\n",
      "Epoch 22/50, Loss: 0.0559\n",
      "Epoch 23/50, Loss: 0.0376\n",
      "Epoch 24/50, Loss: 0.0411\n",
      "Epoch 25/50, Loss: 0.0272\n",
      "Epoch 26/50, Loss: 0.0290\n",
      "Epoch 27/50, Loss: 0.1781\n",
      "Epoch 28/50, Loss: 0.0333\n",
      "Epoch 29/50, Loss: 0.0258\n",
      "Epoch 30/50, Loss: 0.0907\n",
      "Epoch 31/50, Loss: 0.0708\n",
      "Epoch 32/50, Loss: 0.0388\n",
      "Epoch 33/50, Loss: 0.0419\n",
      "Epoch 34/50, Loss: 0.0141\n",
      "Epoch 35/50, Loss: 0.0638\n",
      "Epoch 36/50, Loss: 0.0175\n",
      "Epoch 37/50, Loss: 0.0182\n",
      "Epoch 38/50, Loss: 0.0147\n",
      "Epoch 39/50, Loss: 0.0119\n",
      "Epoch 40/50, Loss: 0.0077\n",
      "Epoch 41/50, Loss: 0.0053\n",
      "Epoch 42/50, Loss: 0.0180\n",
      "Epoch 43/50, Loss: 0.0208\n",
      "Epoch 44/50, Loss: 0.0212\n",
      "Epoch 45/50, Loss: 0.0120\n",
      "Epoch 46/50, Loss: 0.0151\n",
      "Epoch 47/50, Loss: 0.0093\n",
      "Epoch 48/50, Loss: 0.0561\n",
      "Epoch 49/50, Loss: 0.0206\n",
      "Epoch 50/50, Loss: 0.0775\n"
     ]
    }
   ],
   "source": [
    "num_epochs = 50\n",
    "\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "model.to(device)\n",
    "\n",
    "print(f\"\\nTraining on: {device}\")\n",
    "\n",
    "for epoch in range(num_epochs):\n",
    "    model.train()\n",
    "    running_loss = 0.0\n",
    "    for inputs, labels in train_loader:\n",
    "        inputs, labels = inputs.to(device), labels.to(device)\n",
    "\n",
    "        optimizer.zero_grad()\n",
    "        outputs = model(inputs)\n",
    "        loss = criterion(outputs, labels)\n",
    "        loss.backward()\n",
    "        optimizer.step()\n",
    "\n",
    "        running_loss += loss.item() * inputs.size(0)\n",
    "\n",
    "    epoch_loss = running_loss / len(train_dataset)\n",
    "    print(f\"Epoch {epoch+1}/{num_epochs}, Loss: {epoch_loss:.4f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c31df3c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Deep Learning Classification Accuracy on Test Set: 0.8684\n",
      "\n",
      "Deep Learning Classification Report:\n",
      "               precision    recall  f1-score   support\n",
      "\n",
      "       Basal       0.83      1.00      0.91        29\n",
      "        Her2       0.67      0.60      0.63        10\n",
      "        LumA       0.94      0.91      0.93        82\n",
      "        LumB       0.87      0.80      0.83        25\n",
      "      Normal       0.40      0.33      0.36         6\n",
      "\n",
      "    accuracy                           0.87       152\n",
      "   macro avg       0.74      0.73      0.73       152\n",
      "weighted avg       0.87      0.87      0.87       152\n",
      "\n"
     ]
    }
   ],
   "source": [
    "model.eval()\n",
    "correct = 0\n",
    "total = 0\n",
    "all_preds = []\n",
    "all_labels = []\n",
    "\n",
    "with torch.no_grad():\n",
    "    for inputs, labels in test_loader:\n",
    "        inputs, labels = inputs.to(device), labels.to(device)\n",
    "        outputs = model(inputs)\n",
    "        _, predicted = torch.max(outputs.data, 1)\n",
    "        total += labels.size(0)\n",
    "        correct += (predicted == labels).sum().item()\n",
    "\n",
    "        all_preds.extend(predicted.cpu().numpy())\n",
    "        all_labels.extend(labels.cpu().numpy())\n",
    "\n",
    "accuracy_dl = correct / total\n",
    "report_dl = classification_report(all_labels, all_preds, target_names=label_encoder.classes_)\n",
    "\n",
    "print(f\"\\nDeep Learning Classification Accuracy on Test Set: {accuracy_dl:.4f}\")\n",
    "print(\"\\nDeep Learning Classification Report:\\n\", report_dl)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
